### PR TITLE
room DB for local caching

### PR DESCRIPTION
**Pull request by violoncelloCH**
depends on #32 

this is only a start for now
it will serve as a backend for the LocalDataSource of the Repository

room documentation: https://developer.android.com/training/data-storage/room
how to construct relations in room: https://developer.android.com/training/data-storage/room/relationships

### CODE REVIEW

**Comment by violoncelloCH** [comment]
@srsingh04 no, I don't think so... I'm not even sure if it makes sense to have that as an attribute on the data class, but whether a tag has been selected in the UI is state only concerning the UI. It doesn't make sense to store that in DB and especially not on Supabase since that would mean it would get synchronized between all users.

File: app/src/main/java/com/github/swent/echo/data/repository/datasources/Cache.kt
Comment (violoncelloCH) [inline app/src/main/java/com/github/swent/echo/data/repository/datasources/Cache.kt:8]: I'd suggest to rename this to `RoomDataSource` and move it to the room/ dicrectory/package for it to be consistent :)

File: app/src/main/java/com/github/swent/echo/data/room/entity/EventRoom.kt
Comment (violoncelloCH) [inline app/src/main/java/com/github/swent/echo/data/room/entity/EventRoom.kt:15]: If I understand correctly how `@Embedded` works, this will add the fields of `EventCreatorRoom` to the Event table. I'm not sure if this is actually what we want. The event creator should be a subset of the attributes of `userProfile` and then it could be a DB relation between `UserProfile` and `Event`. Otherwise changes to the `UserProfile` won't be visible on the events.

Maybe [views](https://developer.android.com/training/data-storage/room/creating-views) could help us implement this properly? However they only allow to do "SELECT" queries. 
Also the downside of this approach would be that we'd either have a situation where `userProfile`s would be incomplete in the local database (i.e. only the attributes part of `CreatorProfile` have been loaded from the remote). Or we'd need to fetch the complete the compete `UserProfile`s as soon as we have an Event referencing it (not what we want either).

What do you think makes the most sense here? Would the "incomplete `UserProfile`s" approach make sense to you?

File: app/src/main/java/com/github/swent/echo/data/room/entity/UserProfileRoom.kt
Comment (violoncelloCH) [inline app/src/main/java/com/github/swent/echo/data/room/entity/UserProfileRoom.kt:18]: I already started adding association committee memberships as well as association following as relations between `UserProfile` and `Association` over at #157. So this might be something you already want to take a look at depending on which PR gets merged first.

**Review by violoncelloCH — COMMENTED** [review]
Thanks for working on the local DB implementation :) It is really well done in general! 
I have some specific questions and feedback though in the comments below.

File: app/src/main/java/com/github/swent/echo/data/room/entity/EventRoom.kt
Comment (octogradiste) [reply app/src/main/java/com/github/swent/echo/data/room/entity/EventRoom.kt:39]: This is something we should discuss in the team, but as I understood, we want to drop the images as it doesn't add much and adds a lot of complexity. Also, we store an `imageId` (an `int`) inside the `Event`, if we want to use images, we probably should use a `String` to store the path of the image. That why I store a `imagePath` inside the `EventRoom`. But, as I said, this is something we should discuss with the team during the next meeting.

File: app/src/main/java/com/github/swent/echo/data/room/dao/UserProfileRoomDao.kt
Comment (octogradiste) [reply app/src/main/java/com/github/swent/echo/data/room/dao/UserProfileRoomDao.kt:18]: Good catch, I should correct this!

File: app/src/main/java/com/github/swent/echo/data/room/RoomLocalDataSource.kt
Comment (octogradiste) [reply app/src/main/java/com/github/swent/echo/data/room/RoomLocalDataSource.kt:53]: Good point, I'll change that just in case, but as all tests pass (also the one where I set an event with tags and then retrieve it back), I feel like this isn't really an issue when using room.

File: app/src/main/java/com/github/swent/echo/data/room/entity/EventRoom.kt
Comment (octogradiste) [reply app/src/main/java/com/github/swent/echo/data/room/entity/EventRoom.kt:15]: Yeah, you're right, it's not the best solution. My thinking was that the name of the user won't change often. Therefore, I thought it's acceptable that in some rare cases, it might be out of sync with the real user profile name. Also, we are using the local data source only for caching and not as a source of truth. We could add some mechanism in the UI (e.g. pull down to refresh) to update/sync/refresh the data...
Otherwise, we could just not offer the possibility to the user to change his name... this we might also discuss with the team.

File: app/src/main/java/com/github/swent/echo/data/room/entity/UserProfileRoom.kt
Comment (octogradiste) [reply app/src/main/java/com/github/swent/echo/data/room/entity/UserProfileRoom.kt:18]: OK, thanks for the info, I'll take a look!
By the way, thank you so much for taking the time to review all this!!!

File: app/src/main/java/com/github/swent/echo/data/room/entity/EventRoom.kt
Comment (violoncelloCH) [reply app/src/main/java/com/github/swent/echo/data/room/entity/EventRoom.kt:39]: you're right that storing an `int` is not the right thing to do... but I think it makes sense to already prepare this DB column for later once we want to add that feature. So changing that to storing a filepath or some kind of resource identifier probably makes sense... I'll take a look how storing blobs on supabase works.

File: app/src/main/java/com/github/swent/echo/data/room/entity/EventRoom.kt
Comment (violoncelloCH) [reply app/src/main/java/com/github/swent/echo/data/room/entity/EventRoom.kt:15]: Agree that that is a low priority thing and as you say it's not the source of truth it's acceptable to be out of sync. Offering a pull-to-refresh to force-refresh is a good idea! We should look into this.
Treating the names (or similar attributes) as immutable however is generally a bad idea. Users might simply misspell it on profile creation and want to correct it or want to change names for a diverse range of reason (marriage, gender, identity, etc)...

Comment (octogradiste) [reply app/src/main/java/com/github/swent/echo/data/room/entity/EventRoom.kt:15]: Yes, you're right, we should let the user the freedom to change their name. But, as you said, this is not high priority, so we should leave this for the future and keep this solution for now!

File: app/src/main/java/com/github/swent/echo/data/room/entity/EventRoom.kt
Comment (octogradiste) [reply app/src/main/java/com/github/swent/echo/data/room/entity/EventRoom.kt:39]: I create an issue #165 in case we want to add images later on...

File: app/src/main/java/com/github/swent/echo/data/room/entity/EventRoom.kt
Comment (octogradiste) [reply app/src/main/java/com/github/swent/echo/data/room/entity/EventRoom.kt:15]: I opened an issue #166 for the pull-to-refresh support.

**Review by octogradiste — DISMISSED** [review]
@violoncelloCH can't approve because he created the PR, so I have to do it for him ;)

**Comment by octogradiste** [comment]
We might want to wait until #157 is merged. This way, I can adapt the local database to the new changes of the `Event` data class...

**Review by unglazedstamp — APPROVED** [review]
Approved by @violoncelloCH

**Pull request merged**
